node/src/types/chainspec/protocol_config.rs

```diff
@@ -0,0 +1,95 @@
+use datasize::DataSize;
+#[cfg(test)]
+use rand::Rng;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+
+use casper_types::bytesrepr::{self, FromBytes, ToBytes};
+
+use crate::components::consensus::EraId;
+#[cfg(test)]
+use crate::testing::TestRng;
+
+/// The era whose end will trigger the upgrade process.
+#[derive(Copy, Clone, DataSize, PartialEq, Eq, Serialize, Deserialize, Debug)]
+pub(crate) struct ActivationPoint {
+    pub(crate) era_id: EraId,
+}
+
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, DataSize, Debug)]
+pub struct ProtocolConfig {
+    #[data_size(skip)]
+    pub(crate) version: Version,
+    /// This protocol config applies to the era begun immediately after the activation point.
+    pub(crate) activation_point: ActivationPoint,
+}
+
+#[cfg(test)]
+impl ProtocolConfig {
+    /// Generates a random instance using a `TestRng`.
+    pub fn random(rng: &mut TestRng) -> Self {
+        let protocol_version = Version::new(
+            rng.gen_range(0, 10),
+            rng.gen::<u8>() as u64,
+            rng.gen::<u8>() as u64,
+        );
+        let activation_point = ActivationPoint {
+            era_id: EraId(rng.gen::<u8>() as u64),
+        };
+
+        ProtocolConfig {
+            version: protocol_version,
+            activation_point,
+        }
+    }
+}
+
+impl ToBytes for ProtocolConfig {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        buffer.extend(self.version.to_string().to_bytes()?);
```

Comment:
Can we have a `ToBytes/FromBytes` implementations for `Version` instead? It won't be possible to fail with `bytesrepr::Error::Formatting` when parsing it back from `String`.

Reply:
We could, but I actively avoided it here (and in other PRs) because the implementation would have to exist in the casper-types crate, and that crate already provides a similar struct `SemVer` which predates our use of the `semver` crate.

The intent was that we'd eventually replace our own with the `semver` crate, but that crate doesn't support using it with serde in a no-std environment.  Once we have more time, we'll hopefully resolve this problem.